src/binmode/legacy4third.c

```diff
@@ -533,16 +549,66 @@ void legacy_protocol(void) {
                 if (binmode_debug) {
                     printf("\r\n>> ");
                 }
+                bool is_read_sig_cmd = false;  // true when reading any one of the signature bytes
+                uint8_t read_sig_byte_inx = 0; // which signature byte to read
                 for (int i = 0; i < bytes2read; i++) {
                     if (button_get(0)) {
                         return;
                     }
                     if (binmode_debug) {
                         printf("\r\n0x%02X | ", tmpbuf[i]);
                     }
+                    // This hack required to handle parts with more than
+                    // 64K bytes of flash.  Signature byte 2 indicates flash
+                    // size - we'll use this byte to set a global flag if
+                    // this part requires writing the Extended High Byte address
+                    // First - check if we are receiving a command 0x30 (query signature);
+                    // this command is 4 bytes of SPI:
+                    // 0 -> avrdude/BP sends 0x30
+                    // 1 -> avrdude/BP clocks out 0, and part returns 0x30 to confirm command
+                    // 2 -> avrdude/BP sends which of the 3 signature bytes to read
+                    // 3 -> avrdude/BP clocks out 0, part returns requested value.
+                    if (i == 0 && tmpbuf[i] == 0x30) {
+                        // this is the start of a read signature command - set state active
+                        is_read_sig_cmd = true;
+                    }
+                    if (is_read_sig_cmd && i == 2) {
+                        // this is 3rd byte (byte 2) of sequence from BP to part - which signature byte to read
+                        read_sig_byte_inx = tmpbuf[i];
+                    }
                     tmpbuf[i] = hwspi_write_read(tmpbuf[i]);
                     if (binmode_debug) {
                         printf("0x%02X", tmpbuf[i]);
+                        if (is_read_sig_cmd) {
+                            if (i == 0) {
+                                printf("  - read signature command");
+                            } else if (i == 2) {
+                                printf("  - read signature byte %d", read_sig_byte_inx);
+                            }
+                        }
+                    }
+                    // if we are in the process of reading signature command, and
+                    // this is the the 4th byte (byte 3) of the request to read
+                    // signature byte 2, then compare to the known array of
```

Comment:
`signature byte 2` --> `third byte of the signature (address 2)` ... to more closely match datasheet language for `Read Signature Byte command`